### PR TITLE
Electrochemical Nutrient Removal

### PR DESCRIPTION
**Pull request by arao53**
## Fixes/Resolves:
Added flowsheet, sweep analysis, and tests for a single unit electrochemical nutrient removal process. 

## Summary/Motivation:
This analysis uses a zero-order approach to consider levelized cost metrics associated with the conversion of phosphorous and nitrogen to struvite for recovery. 

## Changes proposed in this PR:
- electrochemical_nutrient_removal.py: new, single unit flowsheet 
- multi_sweep.py: sweep uses WaterTAP ParameterSweep tool 
- test_electrochemcial_nutrient_removal.py: component tests for electrochemical nutrient removal flowsheet
- test_multi_sweep.py: tests single run of each parameter sweep (currently only one sweep included)
- costing: unit changes in ZO costing and nanofiltration from currency to currency/time basis. 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+from watertap.core.zero_order_costing import ZeroOrderCosting
+
+
+def main():
+    m = build()
+    set_operating_conditions(m)
+    assert_degrees_of_freedom(m, 0)
+    assert_units_consistent(m)
+
+    solve(m)
```
Comment (tarnold17) [inline watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py]: Recommend changing this to

    results = solve(m)
    assert_optimal_termination(results)

Also for the second instance of `solve(m)` a few lines down.

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    m.fs = FlowsheetBlock(default={"dynamic": False})
+    m.fs.prop = prop_ZO.WaterParameterBlock(
+        default={"solute_list": ["nitrogen", "phosphorus", "struvite"]}
+    )
+
+    # components
+    m.fs.feed = FeedZO(default={"property_package": m.fs.prop})
+    m.fs.product_H2O = Product(default={"property_package": m.fs.prop})
+    m.fs.product_struvite = Product(default={"property_package": m.fs.prop})
+    m.fs.electroNP = ElectroNPZO(
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py:88]: @arao53 Do you recall if pumping costs are included within the electroNP unit? If not, maybe we should add a pump that connects to electroNP.

Comment (arao53) [reply watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py:88]: Pumping costs are not included in the base unit - a pump electricity ZO unit has been added between the feed and electroNP to account for this.

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    m.fs.feed.conc_mass_comp[0, "struvite"].fix(conc_struvite)
+    solve(m.fs.feed)
+
+    # pump
+    m.fs.pump.load_parameters_from_database(use_default_removal=True)
+
+    # electroNP
+    m.fs.electroNP.load_parameters_from_database(use_default_removal=True)
+
+    initialize_system(m)
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py]: For consistency, should you also move this `initialize_system` to main?

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/multi_sweep.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    m = electrochemical_nutrient_removal.main()[0]
+
+    outputs, optimize_kwargs, opt_function = set_up_sensitivity(m)
+
+    sweep_params = {}
+    if case_num == 1:
+        # sensitivity analysis
+        sweep_params["MgCl2_cost"] = LinearSample(
+            m.fs.costing.magnesium_chloride_cost, 135, 538, nx
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/multi_sweep.py:45]: I know we can change this as needed anytime we want, but could you add a comment here describing the specificity behind these values?

**Review by adam-a-a — DISMISSED** [review]
LGTM - any additions to analysis can be made in subseqent PRs as needed.

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py (hunk 3 above)
Comment (arao53) [reply watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/electrochemical_nutrient_removal.py]: Thanks for catching this! With the recent push, it looks like it was called twice consecutively. Now fixed.

**Review by adam-a-a — DISMISSED** [review]

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/multi_sweep.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        )
+    else:
+        raise ValueError(f"{case_num} is not yet implemented")
+
+    output_filename = "sensitivity_" + str(case_num)
```
Comment (tarnold17) [inline watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/multi_sweep.py]: I believe that you want to add `+ ".csv"` to the end of this line.

Comment (arao53) [reply watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/electrochemical_nutrient_removal/multi_sweep.py]: We actually found that it's not necessary to include an extension when using the parameter_sweep tool.

**Review by tarnold17 — APPROVED** [review]

**Review by adam-a-a — APPROVED** [review]

**Pull request merged**